CSS Keyframes names can't be CSS wide keywords

grorg opened the pull request:
#### d71fb1b829e793d50dc0d909865611cdb22e5589
<pre>
CSS Keyframes names can&apos;t be CSS wide keywords
<a href="https://bugs.webkit.org/show_bug.cgi?id=246596">https://bugs.webkit.org/show_bug.cgi?id=246596</a>
rdar://101226253

Reviewed by Tim Nguyen.

The CSS Animations and CSS Values specifications require identifier-based
keyframes names to not be CSS wide keywords, nor &quot;default&quot;, nor &quot;none&quot;.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/keyframes-name-invalid-expected.txt:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeKeyframesRule): Forbid the invalid names.

Canonical link: <a href="https://commits.webkit.org/255640@main">https://commits.webkit.org/255640@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6172654cbf77b4e0557d4c5cbcb4653776a45a3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93151 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2352 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102858 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163155 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97154 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2355 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30679 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85540 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98978 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98818 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1634 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79624 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/28555 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71664 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37068 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34882 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3906 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38753 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40682 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37649 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -651,6 +651,11 @@ RefPtr<StyleRuleKeyframes> CSSParserImpl::consumeKeyframesRule(CSSParserTokenRan
+        // From the CSS Values specification, keyframe names are not allowed to be
+        // CSS wide keywords. Nor can they be "default". And CSS Animations additionally
+        // excludes the "none" keyword.
+        if (CSSParserFastPaths::isUniversalKeyword(name) || equalLettersIgnoringASCIICase(name, "default"_s) || equalLettersIgnoringASCIICase(name, "none"_s))
```
I think this should work and align more closely with spec words (and you won't need to do changes to `isUniversalKeyword`):

```suggestion
        if (isValidCustomIdentifier(nameToken.id()) || nameToken.id() == CSSValueNone)
```

nt1m commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -651,6 +651,11 @@ RefPtr<StyleRuleKeyframes> CSSParserImpl::consumeKeyframesRule(CSSParserTokenRan
         name = nameToken.value().toAtomString();
+        // From the CSS Values specification, keyframe names are not allowed to be
+        // CSS wide keywords. Nor can they be "default". And CSS Animations additionally
+        // excludes the "none" keyword.
```
maybe clarify that this is only when the name is an identifier (it doesn't apply when the name is a string)

grorg replied:
It's in the branch guarded by `if (nameToken.type() == IdentToken) {`

grorg replied:
But I've added more to the comment.

nt1m reviewed (commented):
The commit message should reflect that this is only for names that are identifiers

nt1m commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -651,6 +651,11 @@ RefPtr<StyleRuleKeyframes> CSSParserImpl::consumeKeyframesRule(CSSParserTokenRan
+        // From the CSS Values specification, keyframe names that are identifiers
+        // are not allowed to be CSS wide keywords. Nor can they be "default". And CSS Animations
+        // additionally excludes the "none" keyword.
+        if (isValidCustomIdentifier(nameToken.id()) || nameToken.id() == CSSValueNone)
```
```suggestion
        if (!isValidCustomIdentifier(nameToken.id()) || nameToken.id() == CSSValueNone)
```

The tests are failing because of this.

nt1m replied:
(sorry looks like my original suggestion had the typo)

grorg committed:
CSS Keyframes names can't be CSS wide keywords
https://bugs.webkit.org/show_bug.cgi?id=246596
rdar://101226253

Reviewed by Tim Nguyen.

The CSS Animations and CSS Values specifications require identifier-based
keyframes names to not be CSS wide keywords, nor "default", nor "none".

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/keyframes-name-invalid-expected.txt:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeKeyframesRule): Forbid the invalid names.

Canonical link: https://commits.webkit.org/255640@main
